Game/MyProject/Game/GameStates/BuildTrackState.cpp

```diff
@@ -30,17 +32,106 @@ void BuildTrackState::onExit() {
 }
 
 void BuildTrackState::Update() {
-    if (isNextTriggered())
+    if (isPlayTriggered())
         state_controller->requestChange(SelectCar);
     
+    if (isAddTriggered())
+    {
+        //update the road
+    }
+    
+    if(isRemoveTriggered())
+    {
+        //update the road
+    }
+    
+    if(isSaveTriggered())
+    {
+        //save the road
+    }
+    
+    if(isPreviousTriggered())
+    {
+        //update the road
+    }
+    
+    if(isNextTriggered())
+    {
+        //update the road
+    }
     if (isBackTriggered())
         state_controller->requestChange(Track);
 }
 
 bool BuildTrackState::isNextTriggered() {
-    return IvGame::mGame->mEventHandler->IsKeyDown('n');
+    unsigned int mousex,mousey;
+    if(IvGame::mGame->mEventHandler->IsMousePressed(mousex,mousey))
+    {
+        RayBoxIntersection raybox(state_controller->_main_scene->getCamera()->getRay(mousex,mousey));
```

Comment:
Very redundant code in many game states. You can encapsulate in one single function (defined and implemented in GameState.hpp). Something like 
```
 if(IvGame::mGame->mEventHandler->IsMousePressed(mousex,mousey))
         return ray_intersects_scene_node("node_name",mousex,mousey).
```